### PR TITLE
first pass rc selector

### PR DESCRIPTION
**Pull request by andrewtruong**
Fixes WB-NNNN
Fixes #NNNN

Description
-----------
What does the PR do?

Testing
-------
How was this PR tested?

Checklist
-------
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: wandb/sdk/launch/agent/agent.py [hunk 1]
```diff
@@ -121,25 +126,62 @@ def finish_job_id(self, job_id: Union[str, int]) -> None:
...
+            }
+            """
+        )
+
+        run_queues = self._api.client.execute(QUERY, {"entityName": entity})
+        resource_configs = [json.loads(q["resourceConfig"]) for q in run_queues]
+        valid_resource_configs = [
+            rc for rc in resource_configs if rc["runner"] in self._configured_runners
+        ]
+        return valid_resource_configs
```
Comment (jlzhao27) [inline wandb/sdk/launch/agent/agent.py:165]: can we also support label match for local process and local container runners?

Comment (andrewtruong) [reply wandb/sdk/launch/agent/agent.py:177]: yep

File: wandb/sdk/launch/agent/agent.py [hunk 2]
```diff
@@ -121,25 +127,73 @@ def finish_job_id(self, job_id: Union[str, int]) -> None:
...
 
+    def _get_valid_resource_configs(self, entity: str) -> list[RunQueue]:
+        """
+        Given an entity, return a list of run queues associated with that entity.
+        """
+        QUERY = gql(
+            """
+            query GetRunQueuesByEntity($entityName: String!) {
+                runQueues(entityName: $entityName) {
+                    resource_config
```
Comment (KyleGoyette) [inline wandb/sdk/launch/agent/agent.py:157]: lets also fetch the queue id and name here

Comment (KyleGoyette) [reply wandb/sdk/launch/agent/agent.py:157]: and project

File: wandb/sdk/launch/agent/agent.py [hunk 3]
```diff
@@ -121,25 +127,73 @@ def finish_job_id(self, job_id: Union[str, int]) -> None:
...
+                labels: [ "gpu-pool" ]
+            },
+            local-container: {
+                labels: [ "gpu-pool"]
             }
+        }
         """
         pass
 
+    def _get_valid_resource_configs(self, entity: str) -> list[RunQueue]:
```
Comment (KyleGoyette) [inline wandb/sdk/launch/agent/agent.py:149]: get_valid_queues

File: wandb/sdk/launch/agent/agent.py [hunk 4]
```diff
@@ -121,25 +127,77 @@ def finish_job_id(self, job_id: Union[str, int]) -> None:
...
+        """
+        Given an entity, return a list of run queues associated with that entity.
+        """
+        QUERY = gql(
+            """
+            query GetRunQueuesByEntity($entityName: String!) {
+                runQueues(entityName: $entityName) {
+                    resource_config
+                    queue_id
+                    name
```
Comment (KyleGoyette) [inline wandb/sdk/launch/agent/agent.py:159]: Just a heads up, we'll need the associated project here as well to pop from run queue

File: wandb/sdk/launch/agent/agent.py [hunk 5]
```diff
@@ -121,25 +127,77 @@ def finish_job_id(self, job_id: Union[str, int]) -> None:
...
+            }
+            """
+        )
+
+        run_queues = self._api.client.execute(QUERY, {"entityName": entity})
+        resource_configs = [json.loads(q["resourceConfig"]) for q in run_queues]
+        valid_queues = []
+
+        # can't wait for structural pattern matching...
+        for q, rc in zip(run_queues, resource_configs):
```
Comment (KyleGoyette) [inline wandb/sdk/launch/agent/agent.py:170]: do we need to split this out then rezip it up? coudln't we just check the `resourceConfig` key of the return run_queues here?

File: wandb/sdk/launch/agent/agent.py [hunk 6]
```diff
@@ -121,25 +127,77 @@ def finish_job_id(self, job_id: Union[str, int]) -> None:
...
+            runner = rc["runner"]
+            if runner in self._configured_runners:
+                if runner.startswith("local") and not any(
```
Comment (KyleGoyette) [inline wandb/sdk/launch/agent/agent.py:173]: We should split this into a switch/case for the runner types, then use specific utility functions for each of them. We can add TODOs for the kubernetes , sagemaker and kubernetes runners for now.

File: wandb/sdk/launch/agent/agent.py [hunk 7]
```diff
@@ -121,25 +127,77 @@ def finish_job_id(self, job_id: Union[str, int]) -> None:
...
+        # return random.choice(queues)
+        random.shuffle(queues)
+        return queues
+
+    def _get_combined_config(self):
+        """ "
+        Return a combined config to do the next job.
+        """
+        valid_resource_configs = self._get_valid_queues(self._entity)
+        selected_resource_config = self._select_queue(valid_resource_configs)
```
Comment (KyleGoyette) [inline wandb/sdk/launch/agent/agent.py:194]: I don't think this function will work as expected now. We need to combine the configs in the lists by item.

**Review by KyleGoyette — CHANGES_REQUESTED** [review]
A few comments and requested changes, mostly concerned that the config combination function wont work with combining lists of configs as is.

File: wandb/sdk/launch/agent/agent.py [hunk 8]
```diff
@@ -121,25 +127,77 @@ def finish_job_id(self, job_id: Union[str, int]) -> None:
...
+        QUERY = gql(
+            """
+            query GetRunQueuesByEntity($entityName: String!) {
+                runQueues(entityName: $entityName) {
+                    resource_config
+                    queue_id
+                    name
+                } 
+            }
+            """
```
Comment (dannygoldstein) [inline wandb/sdk/launch/agent/agent.py:162]:
```suggestion
            query GetRunQueuesByEntity($entityName: String!) {
                runQueues(entityName: $entityName) {
                    edges {
                        node {
                              config
                              queue_id
                              name
                         }   
                     }    
                } 
            }
            """
```

**Pull request closed without merging**